gcc/rust/checks/errors/rust-feature-gate.cc

```diff
@@ -39,25 +39,68 @@ FeatureGate::check (AST::Crate &crate)
 		option.parse_to_meta_item ());
 	      for (const auto &item : meta_item->get_items ())
 		{
-		  const auto &name = item->as_string ();
-		  auto tname = Feature::as_name (name);
+		  const auto &name_str = item->as_string ();
+		  auto tname = Feature::as_name (name_str);
 		  if (!tname.is_none ())
-		    valid_features.push_back (Feature::create (tname.get ()));
+		    {
+		      auto name = tname.get ();
+		      valid_features.insert (name);
+		    }
+
 		  else
 		    rust_error_at (item->get_locus (), "unknown feature '%s'",
-				   name.c_str ());
+				   name_str.c_str ());
 		}
 	    }
 	}
     }
-  valid_features.shrink_to_fit ();
 
-  // TODO (mxlol233): add the real feature gate stuff.
   auto &items = crate.items;
   for (auto it = items.begin (); it != items.end (); it++)
     {
       auto &item = *it;
       item->accept_vis (*this);
     }
 }
+
+void
+FeatureGate::gate (Feature::Name name, Location loc,
+		   const std::string error_msg)
+{
+  if (!valid_features.count (name))
+    {
+      auto feature = Feature::create (name);
+      auto issue = feature.issue ();
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-diag"
+      if (issue > 0)
+	rust_error_at (loc,
+		       "%s. see issue %ld "
+		       "<https://github.com/rust-lang/rust/issues/%ld> "
+		       "for more information. add `#![feature(%s)]` "
+		       "to the crate attributes to enable.",
+		       error_msg.c_str (), issue, issue,
+		       feature.as_string ().c_str ());
+      else
+	rust_error_at (loc,
+		       "%s. add `#![feature(%s)]` "
+		       "to the crate attributes to enable.",
+		       error_msg.c_str (), feature.as_string ().c_str ());
+#pragma GCC diagnostic pop
+    }
+}
+
+void
+FeatureGate::visit (AST::ExternBlock &block)
+{
+  if (block.has_abi ())
+    {
+      const auto abi = block.get_abi ();
+
+      if (abi == "rust-intrinsic")
```

Comment:
We have a rust-abi.h so you could do get_abi_from_String () and use an enum instead here.